src/components/Select.tsx

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { motion, MotionStyle, Transition, Variants } from 'framer-motion'
+import { ChevronDown } from 'lucide-react'
+import { ReactNode, useRef, useState } from 'react'
+import styled from 'styled-components'
+
+import useOnOutsideClick from '@/hooks/useOnClickOutside'
+
+interface SelectProps {
+  items: SelectItem[]
+  onItemClick: (value: string) => void
+  selectedItemValue?: string
+  title: string
+  dimensions?: {
+    initialItemHeight?: number
+    expandedItemHeight?: number
+    initialListWidth?: number | string
+    expandedListWidth?: number | string
+  }
+  borderRadius?: number
+  alignText?: 'start' | 'center' | 'end'
+  className?: string
+  style?: MotionStyle
+}
+
+export interface SelectItem {
+  value: string
+  label?: string
+  Component?: ReactNode
+}
+
+const transition: Transition = { type: 'tween', duration: 0.15 }
+
+const Select = ({
+  items,
+  onItemClick,
+  selectedItemValue,
+  title,
+  dimensions = {
+    initialItemHeight: 55,
+    expandedItemHeight: 60,
+    initialListWidth: '98%',
+    expandedListWidth: '100%'
+  },
+  alignText,
+  borderRadius = 12,
+  className,
+  style
+}: SelectProps) => {
+  const [isOpen, setIsOpen] = useState(false)
+
+  const wrapperRef = useRef<HTMLDivElement>(null)
+  const listRef = useRef<HTMLDivElement>(null)
+
+  const { initialItemHeight, expandedItemHeight, initialListWidth, expandedListWidth } = dimensions as Required<
+    typeof dimensions
+  >
+
+  const orderedItems = [
+    ...items.filter(({ value }) => selectedItemValue === value),
+    ...items.filter(({ value }) => selectedItemValue !== value)
+  ]
+
+  const itemContainerVariants: Variants = {
+    initial: {
+      height: initialItemHeight
+    },
+    open: {
+      height: expandedItemHeight,
+      transition
+    }
+  }
+
+  const handleClick = () => {
+    !isOpen && setIsOpen(true)
+
+    // Scroll to top
+    listRef.current?.scroll(0, 0)
+  }
+
+  const handleClickOutside = () => {
+    setIsOpen(false)
+  }
+
+  const handleItemClick = (value: string) => {
+    setIsOpen(false)
+    onItemClick(value)
+  }
```

Comment:
What I learnt recently is that any function that is not wrapped inside a `useCallback` (except the `setState` functions) will get redefined in every component re-render. This has the side-effect that if these functions are passed as props to other components, those components will re-render as well, even if they are memoized (since their props have changed).

So I started wrapping my callback functions that are passed as props with `useCallback`. This ensures that the children components props will change only when the depedencies of the callbacks change, which sounds correct to me.

Reply:
Mhh this need to be evaluated (are the fns complex? Will the trigger a large component re-render?).

In that case I'm not so sure... Maybe the `onItemClick` callback that is received as a prop here should be using `useCallback`..

Some links (also shared in one of you PR):
- https://aheadcreative.co.uk/articles/when-to-use-react-usecallback/
- https://dmitripavlutin.com/react-usecallback/
- https://kentcdodds.com/blog/usememo-and-usecallback

Reply:
Thanks for the articles, very insightful! Indeed, I see now that `useCallback` should be used with care.

What I am referring to is "[Referential equality](https://kentcdodds.com/blog/usememo-and-usecallback#referential-equality)". If a callback is passed from a parent component to a child component, and the child component uses this callback inside a `useEffect` or `useCallback`, then we need to ensure that that effect or callback will not run when the parent component simply re-renders because something inside of it changed, but not the prop callback itself.